Fix part of #14219: Adds test coverage for value generators

jonlake5 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #14219 .
2. This PR does the following: Adds test coverage for value generators. It states missing test on a private class method.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comm
![vmware_usbtkFVdke](https://user-images.githubusercontent.com/5898917/193702961-46e4f70b-9e53-42e8-a58d-66425fea3601.png)
ent "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

jonlake5 committed:
Added test_class_name_of_value_generator

jonlake5 committed:
Added tests for get_html_template

jonlake5 committed:
Added test for refresh_registry

jonlake5 committed:
Update the incomplete coverage text file

jonlake5 committed:
Fix linting issues

jonlake5 committed:
Fix linting issues

jonlake5 committed:
Fix more linting issues

jonlake5 committed:
fixing dirty repo

jonlake5 committed:
Removed unnecessary test and fix line lenghts

jonlake5 committed:
Fixing parentheses

vojtechjelinek commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +48,31 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+        self.assertEqual('BaseValueGenerator', class_object.__class__.__name__)
+
+    def test_registry_template_random_selector_contents(self) -> None:
+        contents_registry = (
+            '<schema-based-editor [schema]="$ctrl.SCHEMA"'
+            ' ng-model="$ctrl.customizationArgs.list_of_values">\n</schema-'
+            'based-editor>\n')
```
```suggestion
        contents_registry = (
            '<schema-based-editor [schema]="$ctrl.SCHEMA" '
            'ng-model="$ctrl.customizationArgs.list_of_values">\n
            '</schema-based-editor>\n'
        )
```

vojtechjelinek commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +48,31 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
                 'generate_value() method has not yet been implemented')):
             base_generator.generate_value()
 
+    def test_class_name_of_value_generator(self) -> None:
+        class_object = value_generators_domain.BaseValueGenerator()
+        self.assertEqual('BaseValueGenerator', class_object.__class__.__name__)
```
I don't see a point of this test.

jonlake5 replied:
I wrote it solely as it stated it didn't have coverage. It can be dropped if you feel it isn't needed.

vojtechjelinek commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +48,31 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+            ' ng-model="$ctrl.customizationArgs.list_of_values">\n</schema-'
+            'based-editor>\n')
+        class_object = value_generators_domain.Registry()
+        self.assertEqual(
+            contents_registry, (class_object.get_generator_class_by_id(
+                'RandomSelector').get_html_template()))
```
```suggestion
        self.assertEqual(
            contents_registry, 
            class_object.get_generator_class_by_id(
                'RandomSelector'
            ).get_html_template()
        )
```

vojtechjelinek commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +48,31 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+                'RandomSelector').get_html_template()))
+
+    def test_registry_template_copier_contents(self) -> None:
+        contents_registry = (
+            '<span class="d-inline-block align-middle">\n  '
+            '<object-editor obj-type="<[objType]>" init-args="initArgs" valu'
+            'e="customizationArgs.value" always-editable="true">\n  </obje'
+            'ct-editor>\n</span>\n')
```
```suggestion
        contents_registry = (
            '<span class="d-inline-block align-middle">\n  '
            '<object-editor obj-type="<[objType]>" init-args="initArgs" '
            'value="customizationArgs.value" always-editable="true">\n  '
            '</object-editor>\n</span>\n'
        )
```

vojtechjelinek commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +48,31 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+            'e="customizationArgs.value" always-editable="true">\n  </obje'
+            'ct-editor>\n</span>\n')
+        class_object = value_generators_domain.Registry()
+        self.assertEqual(
+            contents_registry, class_object.get_generator_class_by_id(
+                'Copier').get_html_template())
+
```
```suggestion
        self.assertEqual(
            contents_registry, 
            class_object.get_generator_class_by_id(
                'Copier'
            ).get_html_template()
        )

```

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

jonlake5 committed:
Update core/domain/value_generators_domain_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

jonlake5 committed:
Update core/domain/value_generators_domain_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

jonlake5 committed:
Update core/domain/value_generators_domain_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

jonlake5 committed:
Update core/domain/value_generators_domain_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

jonlake5 committed:
Merge branch 'develop' into fix-core.domain.value_generators_domain_test

jonlake5 committed:
Removed test that was shown as not necessary by review. Fixed one syntaxt issue from review

jonlake5 committed:
Remove two trailing whitespaces

vojtechjelinek reviewed (approved):
LGTM!

U8NWXD commented:
@jonlake5 This PR looks great! You'll need to fix the coverage issue before it can be merged though. See:

https://github.com/oppia/oppia/actions/runs/3189320683/jobs/5202993655#step:8:1546

jonlake5 commented:
> @jonlake5 This PR looks great! You'll need to fix the coverage issue before it can be merged though. See:
> 
> https://github.com/oppia/oppia/actions/runs/3189320683/jobs/5202993655#step:8:1546

@U8NWXD That line showing no coverage refers to the below protected class method. Line 105 is the one that starts with for _, clazz. I got some differing opinions on testing protected methods. This protected method is called by both get_all_generator_classes and get_generator_class_by_id. Both of these are called in tested methods.

What would be the best way to test this protected method so it shows coverage? 


`
def _refresh_registry(cls) -> None:

        """Refreshes the dictionary mapping between generator_id and the
        corresponding generator classes.
        """

        cls.value_generators_dict.clear()
        # Assemble all generators in
        # extensions/value_generators/models/generators.py.
        module_path_parts = feconf.VALUE_GENERATORS_DIR.split(os.sep)
        module_path_parts.extend(['models', 'generators'])
        module = importlib.import_module('.'.join(module_path_parts))
        for _, clazz in inspect.getmembers(module, predicate=inspect.isclass):
            if issubclass(clazz, BaseValueGenerator):
                cls.value_generators_dict[clazz.__name__] = clazz`

jonlake5 committed:
Merge branch 'develop' into fix-core.domain.value_generators_domain_test

U8NWXD commented:
@jonlake5 you can look at how the protected method is used by unprotected methods and write a test that calls an unprotected method in such a way that the line to be covered in the protected method gets called

jonlake5 commented:
@U8NWXD That makes sense, however, even before starting this PR, the protected method had been called by calling public methods.
I pulled the relevant parts of the value_generators_domain.py and all of the test in the below gist.
https://gist.github.com/jonlake5/0ecd2fc22ab67d68ae312d2ae11bbcae

Lines 16-15 in the gist is what is stating doesn't have coverage. This protected method gets called by both get_all_generator_classes and get_generator_class_by_id.
Those are called in the test on lines 58, 63, 80, and 85.

58 and 63 are ones I added, 80 and 85 is code that was pre-existing

jonlake5 committed:
Added test for test_registry_generator_not_found

jonlake5 committed:
Fix linting

jonlake5 committed:
Fix linting

jonlake5 committed:
Merge branch 'fix-core.domain.value_generators_domain_test' of github.com:jonlake5/oppia into fix-core.domain.value_generators_domain_test
Needed to pull in changes made by oppia devs in public branch

U8NWXD commented:
@jonlake5 even though the existing tests cause the protected method to be called, the incomplete coverage means that there are lines in the protected method that aren't running. Here's a simplified example:

```python
class Score:
    def _floor(self, num):
        if num < 0:
            return 0
        return num
    def get_points(self, score, penalty):
        return self._floor(score - penalty)
```

```python
class ScoreTests
    def test_get_points(self):
        score = Score()
        self.assertEqual(score.get_points(5, 2), 3)
```

In this example, the test calls `get_points()`, which in turn calls `_floor`, but because `5 - 2 > 0`, the `if num < 0` evaluates to False and we never execute the `return 0` line. To cover that line, you'd need to add a test that does something like `score.get_points(5, 6)`. 

The same principle applies to your case. If line 16 in that gist isn't getting called, then that means that `inspect.getmembers(module, predicate=inspect.isclass)` is returning an empty iterable. To get that line covered, I would first look at what `module` is and figure out why `inspect` isn't returning any members for it

jonlake5 commented:
@U8NWXD Thanks for all the clear feedback. I added some print statements to the protected method. It appears from the print statements, that all lines are getting called.

There are/were 2 iterables the whole time. Also the if statement was succeeding and had 2 items being returned. I validated this by also looking at the list that was returned by this.

I added a new gist: https://gist.github.com/jonlake5/1a4fa4c1c4773becb18f0bebba9cd9d7

In the gist 17-13 is equivalent to 108-104 showing a lack of coverage.

In the output portion on lines 28-31 in the gist, it shows all my print statements being called in both the iterated loop as well as inside the if statement (then they are repeated multiple times since this protected method is called multiple times).

Any help or insight is very much appreciated.

seanlip commented:
Assigning @U8NWXD @IamEzio for help with https://github.com/oppia/oppia/pull/16198#issuecomment-1273961194

U8NWXD commented:
@jonlake5 In your [most recent backend test run](https://github.com/oppia/oppia/actions/runs/3213973310/jobs/5254026251), it looks like you're missing branch coverage, not line coverage

```
 INCOMPLETE COVERAGE (97.0%): core.domain.value_generators_domain_test
Name                                     Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------------------
core/domain/value_generators_domain.py      33      0      6      1    97%   106->105
------------------------------------------------------------------------------------
TOTAL                                       33      0      6      1    97%
```

Notice how the number of `Miss` statements is 0, but you have one partial branch (`BrPart`). That branch is the jump from line 106 to line 105. Here's the relevant code:

https://github.com/oppia/oppia/blob/b9d1b9d3040ed03593d4afa7d5d97717796f127b/core/domain/value_generators_domain.py#L105-L107

The branch 106 -> 105 happens when `issubclass(clazz, BaseValueGenerator)` is False, so I suspect that there are always two iterables, both of which are subclasses of `BaseValueGenerator`, so the tests never explore the branch where `clazz` is not a subclass of `BaseValueGenerator`. Does this check out with your analysis?

IamEzio commented:
Thanks @U8NWXD. @jonlake5 I think covering the branch `106 -> 105` as stated in the comment above would solve the problem.

jonlake5 commented:
@U8NWXD @IamEzio That makes sense and thanks for the info on how to read the coverage. That does explain my analysis of the tests. I would have to have a file that isn't a subclass of BaseValueGenerator for that branch to happen.

jonlake5 committed:
Added an empty class to complete branch testing

jonlake5 committed:
Fix formatting issues

jonlake5 committed:
Merge branch 'develop' into fix-core.domain.value_generators_domain_test

jonlake5 commented:
@U8NWXD I added the below to core.domain.extensions.value_generators.generators and now have 100% coverage.



```
class TestValueGeneratorCodeBranch():
    """For complete branch testing coverage there
    needs to be a class in here that is not a subclass
    of BaseValueGenerator.
    """

    pass
```

seanlip commented:
@jonlake5 Let's not add dead code to the main code just for the sake of tests. That sounds like the tail wagging the dog.

Instead, in your test file, can you use self.swap (see test_utils) to swap out one of the classes and replace it with a mock class that is not a subclass of BaseValueGenerator? That would achieve your objective, I think, while still keeping unneeded code out of what is served to the user.

seanlip reviewed (changes requested):
Leaving a review hold (see previous comment).

jonlake5 committed:
remove dead function to use self.swap in test

jonlake5 committed:
Added testing using self.swap to hit all code branches

jonlake5 committed:
Fix formatting issues

jonlake5 commented:
@seanlip Thanks for that guidance. I have reverted changes made previously and updated the test file as recommended. It should be good for review.

seanlip commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +62,67 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+        )
+
+
+class NotBvgSubclass():
```
Define this class within your test function, so that it's scoped only to that test and not available to other tests.

jonlake5 replied:
Done.

seanlip commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +62,67 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+
+class NotBvgSubclass():
+
+    """This is a dummy class for self.swap to test all code branches.
```
The aim of a test should never be "to test all code branches".

In this case, what you're trying to test is that the value generator registry discovers all classes correctly and excludes classes that are not subclasses of BaseValueGenerator. Adding this context is important for future readers of the code.

jonlake5 replied:
Done

seanlip commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +62,67 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+        module_path_parts = feconf.VALUE_GENERATORS_DIR.split(os.sep)
+        module_path_parts.extend(['models', 'generators'])
+        module = importlib.import_module('.'.join(module_path_parts))
+        vg_class_names = list(
```
You know what the list is already since the file is in the codebase, so just swap out one of the existing value generators with your new mock class (which you can e.g. name MockCopier). It's fine to hardcode the file name etc. rather than try to discover it programmatically here.

jonlake5 replied:
Done

seanlip commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +62,67 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+        vg_class_names = list(
+            value_generators_domain.Registry.get_all_generator_classes().keys()
+        )
+        if len(vg_class_names) > 0:
```
This is too complex -- don't keep logic in tests.

Instead, after the swap, do the get() and then do a self.assertEqual on the result of the get(). That is sufficient -- if someone changes the code then they'll change the test, and this code doesn't change frequently.

jonlake5 replied:
Done

seanlip reviewed (changes requested):
Thanks @jonlake5! Took a follow-up pass.

jonlake5 committed:
Merge branch 'develop' into fix-core.domain.value_generators_domain_test

jonlake5 committed:
Update tests to simplify logic and test setup

jonlake5 committed:
Merge branch 'fix-core.domain.value_generators_domain_test' of github.com:jonlake5/oppia into fix-core.domain.value_generators_domain_test

jonlake5 commented:
@seanlip Thanks for the review and providing that insight. I made the changes.

seanlip commented:
Thanks @jonlake5 -- did you forget to respond to the review comments though? Please follow the [PR process](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve) here and respond to all comments before reassigning reviewers -- thanks.

jonlake5 commented:
@seanlip PTAL

seanlip commented on core/domain/value_generators_domain_test.py:
```diff
@@ -48,6 +61,63 @@ def test_generate_value_of_base_value_generator_raises_error(self) -> None:
+        )
+        expected_generators = {
+            'RandomSelector': type(generators.RandomSelector())
+            }
```
Deindent by 4 to match opening {.

jonlake5 replied:
Done

seanlip replied:
Er, not done?

Please push your code changes and only then reply to the comments, so that the code is consistent with your comments. (In general, reviewers get notified once comments are replied to, so by that time the code should be in a reviewable state.) Thanks!

jonlake5 replied:
Sorry about that. I went back to my VM and it was hung after the type checks finished but never pushed the changes. I pushed again and checked the branch on github and it looks good now.

seanlip reviewed (approved):
Looks good, just one small thing. Thanks!

seanlip commented:
Also congrats on your first PR to Oppia! :)

U8NWXD commented:
Congrats @jonlake5! Can you please assign me once you address Sean's last comment so I can help get this merged?

jonlake5 committed:
Update formatting

jonlake5 commented:
@U8NWXD PTAL

seanlip reviewed (changes requested):
Updating assignments and leaving a review hold this time. There's still a fix needed, so this should be assigned to @jonlake5.

seanlip reviewed (approved):
Awesome, thanks @jonlake5!

Ah -- for future reference, you can double-check the "Files changed" tab on GitHub just to make sure that it reflects your changes. That's the most foolproof way since it's the same as what the reviewers would see. 

Hope that helps! Congrats again on your first PR :)

jonlake5 commented:
> Awesome, thanks @jonlake5!
> 
> Ah -- for future reference, you can double-check the "Files changed" tab on GitHub just to make sure that it reflects your changes. That's the most foolproof way since it's the same as what the reviewers would see.
> 
> Hope that helps! Congrats again on your first PR :)

Thanks I appreciate it. Appreciate the help and patience as I work through my first PR here (with hopefully more to come).

seanlip commented:
Ah, sorry, @jonlake5 -- our Python version was upgraded just before your PR got merged :( So now the last CI check can't run.

Would you mind following the instructions [here](https://groups.google.com/g/oppia-dev/c/Ytvepw_TfpQ), please, and merging develop into your branch as well? Thanks!

jonlake5 committed:
Merge branch 'develop' into fix-core.domain.value_generators_domain_test

jonlake5 committed:
Updated schema contents in test as it changed since test was written

jonlake5 commented:
@seanlip I updated python and merged the changes into my branch (at least I think I did it right)

seanlip commented:
Looks great, merging. Thanks @jonlake5!!
